ROVER-257 Hot-reloading of Federation Version in `supergraph.yaml`

jonathanrainer opened the pull request:
One of the key features that is required for the LSP is the ability to change the Federation Version specified in the `supergraph.yaml` and for it to then reload in the active session, rather than needing a restart. This is implemented as of this PR, and due to the nature of the previous refactor, has also been added to `rover dev` as well!

The way this has been done is to extend the FederationWatcher to track when a change of version occurs and then to get the CompositionWatcher to update its own internal state (downloading a new version when necessary). The outcome has been tested and performs w.r.t the LSP as we'd expect.  Below is an example of the logging with it working for `rover dev`:

![Screenshot 2025-01-09 at 12 06 16](https://github.com/user-attachments/assets/b961054f-6730-4770-a380-f25edeaacdaf)

I've also implemented a feature whereby we don't do the auto-updating behaviour if a user is specifically overriding the version of composition to be used via an environment variable, though for the moment this is only relevant to `rover dev`. 

There is an important open question here w.r.t behaviour:
- The original ticket mentioned a behaviour where if the change didn't work the router would shut down, but this seems a bit overmuch. A much better solution from a UX perspective, and the one implemented here, is to continue composing with the old version but raise an error so the user can correct it. I am happy to hear the counter argument on this though!

jonathanrainer committed:
ROVER-257 Add events for when Federation Version changes

Now we are tracking changes in the Federation Version explicitly,
so we should be able to see when a change occurs.

jonathanrainer committed:
ROVER-257 Fix tests

jonathanrainer committed:
ROVER-257 Ensure we don't compose on error

Further make the error message a bit more verbose

jonathanrainer committed:
ROVER-257 Make nicer user focussed error messages

jonathanrainer committed:
ROVER-257 Don't provide update behaviour if overrides are in place

If the user has specified an env var/CLI flag to override the federation
version, then don't let updates ruin that. Rather log out that a change
should be happening, but isn't due to the override.

jonathanrainer committed:
ROVER-257 Ensure we don't re-compose on errors

aaronArinder commented on src/composition/watchers/watcher/supergraph_config.rs:
```diff
@@ -184,10 +190,23 @@ impl SupergraphConfigDiff {
+                "Old: {:?}, New: {:?}",
+                old.get_federation_version(),
+                new.get_federation_version()
+            );
```
do you want to keep this? if so, mind making it a bit more descriptive?

aaronArinder reviewed (approved):
lgtm!

jonathanrainer committed:
ROVER-257 Make log lines clearer

jonathanrainer committed:
ROVER-257 Action the detected change

Now that we can properly detect the change we enforce the correct
action to occur on top of it, i.e. the federation version changes
as we would anticipate.

jonathanrainer committed:
ROVER-257 Add better logging for `rover dev`
